README.md
I really like using HTML5 boilerplate as the basis for all my websites, including my Django projects. In the past, I have included it as part of other projects, but find myself repeating the same steps for each new project. The Django HTML5 Boilerplate (DH5BP) project addresses this issue by incorporating the HTML5 Boilerplate (H5BP) project into an easy to consume Django-friendly Python package.

Find out for about HTML5 Boilerplate at https://github.com/h5bp/html5-boilerplate

This project differentiates itself from other H5BP to Django ports, by including Fabric scripts that automatically convert new versions of H5BP into a Django-friendly Python module. Ideally, this will make keeping up-to-date with the latest version of H5BP trivial.

Installation
============

Code is found at::

> https://github.com/mattsnider/django-html5-boilerplate

The easiest way to install is using pip::

> pip install django-html5-boilerplate

Requirements
============

To consume the package, you need only have a version of Django >= 1.3. This is the only dependency in the setup file.

To update the repository with the latest H5BP using the Fabric script, you will also need to install Fabric. I use Fabric >= 1.7, but the script is still very simple, so it will probably work on Fabric >= 1.0.

This library has been tested on Python >= 2.6.

Usage
=====

All static files and templates are namespaced under the directory DH5BP. You will need to include DH5BP in your `settings.py`:

    INSTALLED_APPS = (
        'django.contrib.auth',
        'django.contrib.contenttypes',
        'django.contrib.sessions',
        'django.contrib.sites',
        'django.contrib.messages',
        'django.contrib.staticfiles',
        # Uncomment the next line to enable the admin:
        'django.contrib.admin',
        # Uncomment the next line to enable admin documentation:
        'django.contrib.admindocs',
        ...
        'dh5bp',
        ...
    )

H5BP Template
-------------
Any template you want to inherit the H5BP page architecture include the following:

    {% extends 'dh5bp/base.html' %}
    {% load url from future %}
    {% load staticfiles %}
    {% block title %}YOUR TITLE HERE{% endblock %}
    {% block content %}YOUR PAGE MARKUP HERE{% endblock %}

I've added a second template that puts all the JavaScript includes in the `head` tag. I don't recommend using this, unless you are using a mobile library like jQuery Mobile, which expects to be loaded before the body.

    {% extends 'dh5bp/base_script_in_head.html' %}
    {% load url from future %}
    {% load staticfiles %}
    {% block title %}YOUR TITLE HERE{% endblock %}
    {% block content %}YOUR PAGE MARKUP HERE{% endblock %}

Additionally, you can define more styles, meta, and other head-related markup in the title block:

    {% block head %}
        <link rel="stylesheet" href="{% static "css/YOUR_CSS.css" %}">
        <meta name="keywords" content="YOUR KEYWORD">
        ...
    {% endblock %}

The project expects you to create a /static/js/main.js file (see JavaScript notes below) for your primary script. To include JavaScript before or after main.js:

    {% block pre_main_script %}
        <script>
            var MySite = {
                // maybe put global variables passed from server here
            };
        </script>
        <!-- maybe another script file, like require.js -->
    {% endblock %}

    {% block post_main_script %}
        <script src="{% static "js/OTHER_JAVASCRIPT_FILE.js" %}"></script>
    {% endblock %}

If the enduser is viewing your site on a version of IE <= 7, then they will be shown a message indicating that they are using an outdated browser. To overwrite that message:

    {% block outdated_message %}
        <p class="browsehappy">YOUR MESSAGE</p>
    {% endblock %}

In v1.0.3+, there are two new template variables `html_attr` and `body_attr`, which can be used to add attributes to those html elements (such as `id` to the `body` tag and `manifest` to `html` tag). These variables should be provided as strings to the template.

If you use Google Analytics (GA), then H5BP includes the script for it right into the page. You will need to provide the template variable `DH5BP_GA_CODE` with your GA code to activate the script. You will probably want to use a context_processor for this:

> https://docs.djangoproject.com/en/dev/ref/templates/api/#writing-your-own-context-processors

JavaScript
----------
You will be provided the latest `jQuery`, `modernizr`, and a `console` polyfill from H5BP. In addition, the base template will look for a static file `js/main.js`. Put any JavaScript that is required for all pages of your site in `js/main.js`. I have created two blocks (as described above), where you can put JavaScript that needs to be execute before and after the main script.

If you want to use a different file name, instead of `js/main.js`, then provided the template variable `main_js` with the relative path to your JavaScript file from `STATIC_ROOT`. This variable should be provided as strings to the template.

Urls & Views
------------
The H5BP 404 page was ported over and wired up, as well as a similar looking 500 page. To use these in your project add the following to your `urls.py`:

    handler404 = 'dh5bp.views.page_not_found'
    handler500 = 'dh5bp.views.server_error'

I have wired up the default `favicon.ico`, `apple-touch-icon.png`, `humans.txt`, `robots.txt`, and `crossdomain.xml` from H5BP as well. To include those into your project, simply append the DH5BP urls to your urls:

    from dh5bp.urls import urlpatterns as dh5bp_urls

    urlpatterns = patterns('',
        # YOUR URLS
    )
    urlpatterns += dh5bp_urls

If you choose to change the behavior of these URLs, you can either not include these urls or declare your own version before adding the DH5BP urls (whichever definition occurs first will be the one used by Django).

What Isn't Included
===================

H5BP includes an `.htaccess` file for use with apache. Much of what this file is doing is outside of the scope of Django and should be handled by whatever static fileserver you are using, so I did not include it in this project.

Roadmap
=======

I don't think there is much missing right now, but I would like the to make the Fabric scripts more robust and to fail loudly if some step doesn't execute correctly, so that we know that something big has changed in H5BP.

Issues
======

https://github.com/mattsnider/django-html5-boilerplate/issues

Licensing
=========

Apache 2.0; see LICENSE file


CHANGES.txt
v1.0.8, 10/28/2013 -- updating to latest changes of h5bp: normalize v2.1.3, google search removed from error pages, other minor changes. Also, fixing fabric script to insert the blocks introduced in v1.0.7.

v1.0.7, 10/20/2013 -- adding a couple of blocks to allow easier overriding of jquery, modernizr, and normalization

v1.0.6, 9/18/2013 -- updating to latest changes of h5bp (nothing major)

v1.0.5, 8/21/2013 --
Cleaning up fabfile, using more variables, so named blocks and template variables are easy to identify
Adding a second base.html ('base_script_in_head.html') that moves the script blocks to head for use with mobile libraries like jQuery

v1.0.4, 8/20/2013 --
Making `html_attr` and `body_attr` safe, so they render with &quot;.
Cleaning up the fabfile, so you can specify whether to purge the h5bp repo or not.
Allowing consumer to define a new name for "main.js" using `main_js` template variable.

v1.0.3, 8/20/2013 -- The `html_tag` solution does not work, so moving to a template variable {{ html_attr }}. Also adding the {{ body_attr }} template variable.

v1.0.2, 8/20/2013 -- DON'T USE THIS VERSION (`html_tag` BUGGED). Icons updated again and adding a new block `html_tag` in base.html, so you can add things like a manifest file.

v1.0.1, 8/15/2013 -- Incorporating commit cd981e52f7cb96260f9703002885db154a10ebfb, updating the apple icons.

v1.0.0, 8/13/2013 -- Initial release.


fabfile.py
import os, re
from fabric.api import local


def cleanPackage():
    """
    Remove purges the package files that were created from H5BP.
    """
    local('rm -rf dh5bp/static')
    local('rm -rf dh5bp/templates')


def cleanup_h5bp():
    """
    Remove the fetched H5BP.
    """
    local('rm -rf html5-boilerplate')


def getLatestH5BP(cleanup=True):
    """
    Fetch the latest H5BP.
    """
    path_exists = os.path.exists('./html5-boilerplate')

    if cleanup and path_exists:
        cleanup_h5bp()
        path_exists = False

    if not path_exists:
        local('git clone https://github.com/h5bp/html5-boilerplate.git')


def migrateHtml():
    """
    Creates the necessary HTML template files and updates them as necessary to
    make them Django templates.
    """
    e404PathSource = os.path.join('html5-boilerplate', '404.html')
    indexPathSource = os.path.join('html5-boilerplate', 'index.html')

    templatePath = os.path.join('dh5bp', 'templates', 'dh5bp')
    local('mkdir -p %s' % templatePath)

    e404PathDest = os.path.join(templatePath, '404.html')
    e500PathDest = os.path.join(templatePath, '500.html')
    basePathDest = os.path.join(templatePath, 'base.html')
    baseJsOnTopPathDest = os.path.join(templatePath, 'base_script_in_head.html')

    local('cp %s %s' % (e404PathSource, e404PathDest))

    var_body_attr = 'body_attr'
    var_ga_code = 'DH5BP_GA_CODE'
    var_html_attr = 'html_attr'
    var_main_js = 'main_js'

    block_content = 'content'
    block_description = 'description'
    block_head = 'head'
    block_jquery = 'jquery'
    block_modernizr = 'modernizr'
    block_normalize = 'normalizations'
    block_outdated_message = 'outdated_message'
    block_post_main_script = 'post_main_script'
    block_pre_main_script = 'pre_main_script'
    block_title = 'title'
    
    ind_nl = '\n        ';

    # create a 500 error page
    with open(e404PathSource, 'r') as f:
        text = ''.join(f.readlines())
        text = text.replace('Page Not Found', 'Internal Error')
        text = text.replace('Not found', 'Internal Error')
        text = re.sub('<p>[^<]+</p>', '', text)
        text = re.sub(re.compile(
            '<ul>.*?</ul>', flags=re.MULTILINE|re.DOTALL), '', text)
        text = text.replace(
            '</h1>', '</h1>\n'
                     '<p>We are sorry for the inconvenience, '
                     'but something went wrong.</p>'
                     '<p>We have logged the problem and will be '
                     'investigating soon.</p>'
                     '<p>Please go back and try again.</p>')

    with open(e500PathDest, 'w') as f:
        f.write(text)

    # prepare index.html for being a base template
    with open(indexPathSource, 'r') as f:
        text = ''.join(f.readlines())

        # add load staticfiles so we can use the static template tags
        text = text.replace(
            '<!DOCTYPE html>', '{% load staticfiles %}<!DOCTYPE html>')

        # remove non-conditional HTML comments
        text = re.sub('<!--(?!\[|>|<).+-->', '', text)

        # add a attribute template variables to the HTML and BODY tag
        text = text.replace(
            '<html class', '<html {{ %s|safe }} class' % var_html_attr)
        text = text.replace(
            '<body>', '<body {{ %s|safe }}>' % var_body_attr)

        # update the head block (title, meta, additional scripts)
        text = text.replace(
            '<title></title>',
            '<title>{%% block %s %%}{%% endblock %%}</title>' % block_title)
        text = text.replace(
            'content=""',
            'content="{%% block %s %%}{%% endblock %%}"' % block_description)
        text = text.replace(
            '</head>',
            '    {%% block %s %%}{%% endblock %%}\n'
            '    </head>' % block_head)
        text = re.sub('(<link[^>]+normalize\.css)', '{%% block %s %%}%s\\1' % (
        	block_normalize, ind_nl), text)
        text = re.sub('(main\.css[^>]+>)', '\\1%s{%% endblock %%}%s{%% block %s %%}' % (
        	ind_nl, ind_nl, block_modernizr), text)
        text = re.sub('(%s-.*?<\/script>)' % block_modernizr, '\\1%s{%% endblock %%}' % 
        	ind_nl, text)
        text = re.sub('(<script[^>]+%s\.min\.js)' % block_jquery, 
        	'{%% block %s %%}%s\\1' % (block_jquery, ind_nl), text)
        text = re.sub('(plugins\.js.*?<\/script>)',
        	'\\1%s{%% endblock %%}' % ind_nl, text)


        # make CSS and JS paths use static
        text = re.sub('css/([^"]+)', "{% static 'css/dh5bp/\\1' %}", text)
        text = re.sub('js/([^"]+)', "{% static 'js/dh5bp/\\1' %}", text)

        # setup a GA variable, only include script block in final page,
        #   if the code is set
        text = text.replace('UA-XXXXX-X', '{{ %s }}' % var_ga_code)
        text = text.replace(
            '<script>\n', '{%% if %s %%}<script>\n' % var_ga_code)
        text = text.replace(
            '%s</script>' % ind_nl, '%s</script>{%% endif %%}' % ind_nl)

        # create the content block
        text = text.replace('<p>', '{%% block %s %%}<p>' % block_content)
        text = text.replace('</p>', '</p>{% endblock %}')

        # wrap the legacy browser message, so it can be changed
        text = text.replace(
            '<p class="browsehappy">',
            '{%% block %s %%}<p class="browsehappy">' % block_outdated_message)

        # cleanup extra lines
        rx_empty_line_matcher = re.compile(
            '((\s+)\n)+')
        text = re.sub(rx_empty_line_matcher, '\n', text)

        # grab and remove the bottom script section
        rx_script_matcher = re.compile(
            '(<script src=".*main\.js.*?<\/script>)')
        script_section = re.search(rx_script_matcher, text).group(0)
        script_in_head = re.sub(rx_script_matcher, '', text)
        replacement_text = """
        {%% block %s %%}{%% endblock %%}
        <script src="{%% if %s %%}{%% static %s %%}{%% else %%}{%% static 'js/main.js' %%}{%% endif %%}"></script>
        {%% block %s %%}{%% endblock %%}""" % (
            block_pre_main_script, var_main_js, var_main_js,
            block_post_main_script)
        text = re.sub(rx_script_matcher, '', text)
        script_section = '%s%s' % (
            script_section, replacement_text)
        script_section = script_section.replace(
            '<script src="{% static \'js/dh5bp/main.js\' %}"></script>', '')
        text = text.replace(
            '{% if DH5BP_GA_CODE %}',
            '%s%s{%% if DH5BP_GA_CODE %%}' % (script_section, ind_nl))

        # create a copy with scripts in the head section
        rx_jquery_matcher = re.compile(
        	"(\n\s+{% block jquery %}.*?{% endblock %}\n)", flags=re.MULTILINE|re.DOTALL);
        script_section = '%s%s' % (
        	re.search(rx_jquery_matcher, script_in_head).group(0), script_section);
        script_in_head = re.sub(rx_jquery_matcher, '',script_in_head);
        script_in_head = script_in_head.replace('</head>', """    %s
    </head>""" % script_section)

    # write the new text to base.html
    with open(basePathDest, 'w') as f:
        f.write(text)

    # write the new text to base_script_in_head.html
    with open(baseJsOnTopPathDest, 'w') as f:
        f.write(script_in_head)

    # Other Templates
    local('cp %s %s' % (
        os.path.join('html5-boilerplate', "*.xml"), templatePath))
    local('cp %s %s' % (
        os.path.join('html5-boilerplate', "*.txt"), templatePath))

def moveStatic():
    """
    Copy static files from H5BP to the appropriate place in the module.
    """
    # CSS
    cssPathSource = os.path.join('html5-boilerplate', 'css')
    cssPathDest = os.path.join('dh5bp', 'static', 'css', 'dh5bp')
    local('mkdir -p %s' % cssPathDest)
    local('cp %s %s' % (os.path.join(cssPathSource, "*.css"), cssPathDest))

    # JS
    jsPathSource = os.path.join('html5-boilerplate', 'js')
    jsPathDest = os.path.join('dh5bp', 'static', 'js', 'dh5bp')
    local('mkdir -p %s' % jsPathDest)
    local('cp %s %s' % (os.path.join(jsPathSource, "*.js"), jsPathDest))
    local('cp -R %s %s' % (os.path.join(jsPathSource, "vendor"), jsPathDest))

    # Images
    imgPathDest = os.path.join('dh5bp', 'static', 'img', 'dh5bp')
    local('mkdir -p %s' % imgPathDest)
    local('cp %s %s' % (
        os.path.join('html5-boilerplate', '*.png'), imgPathDest))
    local('cp %s %s' % (
        os.path.join('html5-boilerplate', '*.ico'), imgPathDest))


def update(purge=False, cleanup=True):
    """
    Run everything.
    Use "fab update:cleanup=" for setting cleanup to False
    """
    if purge:
        cleanPackage()
    getLatestH5BP(cleanup)
    moveStatic()
    migrateHtml()
    if cleanup:
        cleanup_h5bp()